### ISSUE TITLE
Wrong detecetion of the kernel vesion if the file name include more that one '-' character

### ISSUE THREAD

**Issue by DidierSpaier** [Antynea/grub-btrfs#242]
Hello,

I maintain the Slint  Linux distribution: https://slint.fr. We use preferably btrfs and ship  timeshift and I would  like to provide an easy way to boot off a snapshot using  grub-btrfs, at least on demand. I have packaged  version 4.12. While testing at first no boot entry was generated. This is because our kernel have a naming scheme like vmlinuz-generic-6.0.8 so computing the version with:
kversion=${k#*"-"}  gives:  generic-6.0.8
Writing instead:  kversion=${k##*"-"}  gives 6.0.8 which allows to generate the boot entries.

Best,
Didier

**Comment by Schievel1** [comment]
Idk probably gonna use something more reliable like some sed magic. 
Otherwise the next one comes up with a kernel name with 4 times `-` and things fall apart.

**Comment by DidierSpaier** [comment]
It works for `4` times -  as sown below:
name="one-two-three-four-6.0.8"
`echo ${name##*-}`
6.0.8
Incidentally the single quotes surrounding `-` are not necessary.

Reference: https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_06_02

**Comment by Schievel1** [comment]
I see what you did there...

**Comment by Antynea** [comment]
hello,

In the absence of a naming convention, `Grub-btrfs` is based on the naming scheme of popular distributions.
There must be a mistake in the way you name your initramfs.
`generic` isn't optional and is part of the kernel version used.

```
vmlinuz-generic-6.x.x
initramfs-generic-6.x.x.img
```

If you use a alternative kernel such as , linux hardened , linux lts, etc : 
```
vmlinuz-generic-6.x.x
vmlinuz-hardened-6.x.x
vmlinuz-lts-6.x.x
```
```
initramfs-generic-6.x.x.img
initramfs-hardened-6.x.x.img
initramfs-lts-6.x.x.img
```
or for an alternative initramfs:
```
initramfs-generic-6.x.x-fallback.img
initramfs-hardened-6.x.x-fallback.img
initramfs-lts-6.x.x-fallback.img
```

If we don't respect these few rules, the association of the initramfs with their respective kernels would be chaotic.

**Comment by Schievel1** [comment]
But his kernel name *is* `vmlinuz-generic-6.0.8`.
I don't get it. The reason it is not working must be something different then.

**Comment by DidierSpaier** [comment]
Hello and thanks both for responding.

The only kernel shipped in Slint 15.0 and its associated initramfs generated by dracut are currently named :
`vmlinuz-generic-6.0.8`
`initramfs-6.0.8`

Whether this be a mistake  is debatable, in absence of a specification.

What matters to me is that grub-mkconfig and os-prober  be able to generate boot entries with this naming scheme, and they do. As an aside, the name `vmlinuz-generic-6.0.8`  inherits from Slackware's  (of which Slint is a derivative) naming, but as we do not ship as Slackware does a `huge` kernel, meaning  all included and without an initramfs, I could just use the name `vmlinuz-6.0.8` in the future. As it is now writing `kversion=${k##*"-"}` is good enough for Slint, but maybe does not fit other naming schemes, so please consider my initial message as just an information.

Anyway I will need to patch the script 41_snapshots-btrfs for Slint, to provide a fully flat layout of the grub menu  (no sub-menu). This is because Slint aims to be fully accessible to the blind and this allows to navigate in the menu  using only the up and down arrow keys (users can have an accessible preview of the layout of the GRUB menu before a reboot, running grub-emu or the script https://github.com/DidierSpaier/slint-scripts/blob/master/list_boot_entries (only for a flat layout as-is).

The target use case is to just allow users to make a snapshot using Timeshift before a possibly risky modification of their system and use grub-btrfs to associate to it a boot entry, without further automation for now (maybe in the future), nothing more fancy. Obviously grub-btrfs is a good fit for that, thanks a bunch!

Bonne soirée/Have a good day.

**Comment by DidierSpaier** [comment]
> But his kernel name _is_ `vmlinuz-generic-6.0.8`. I don't get it. The reason it is not working must be something different then.

This could be because the initramfs is named `initramfs-6.0.8` instead of `initramfs-generic-6.0.8` see my other post below (which I sent before seeing yours).

**Comment by Schievel1** [comment]
Yes, that is the actual issue here. We can not associate the initramfs with the kernel.
